ambry-commons/src/main/java/com/github/ambry/commons/PipedAsyncWritableChannel.java

```diff
@@ -0,0 +1,344 @@
+     * Invokes the final callback and updates the future of registered {@link AsyncWritableChannel}.
+     * @param exception the {@link Exception}, if any, to pass to the callback.
+     */
+    private void invokeFinalCallback(Exception exception) {
+      if (callbackInvoked.compareAndSet(false, true)) {
+        if (writableChannel != null) {
+          writableChannelFutureResult.done(totalBytesRead.get(), exception);
+          if (writableChannelFinalCallback != null) {
+            writableChannelFinalCallback.onCompletion(totalBytesRead.get(), exception);
+          }
+        }
+      }
+    }
+
+    @Override
+    public boolean isOpen() {
+      return channelOpen.get();
+    }
+
+    @Override
+    public void close() {
+      lock.lock();
+      try {
+        if (channelOpen.compareAndSet(true, false)) {
+          ChunkData chunkData;
+          while ((chunkData = bufferedChunks.poll()) != null) {
+            chunkData.resolveChunk(new ClosedChannelException());
+          }
+        }
+      } finally {
+        lock.unlock();
+      }
+    }
+  }
+
+  /**
+   * Representation of all the data associated with a chunk i.e. the actual bytes and the future and callback that need to
+   * be invoked on resolution.
+   */
+  private class ChunkData {
+    // The future that will be set on chunk resolution
+    public final FutureResult<Long> future = new FutureResult<>();
+    public ByteBuf buf;
+    private final Callback<Long> callback;
+    private final long size;
+    private Result primaryWriteCallbackResult;
+    private Result secondaryWriteCallbackResult;
+    private final Lock lock = new ReentrantLock();
+
+    /**
+     * Create a new instance of ChunkData with the given parameters.
+     * @param buf            the bytes of data associated with the chunk.
+     * @param callback       the {@link Callback} that will be invoked on chunk resolution.
+     */
+    private ChunkData(ByteBuf buf, Callback<Long> callback) {
+      this.buf = buf;
+      if (buf != null) {
+        size = buf.readableBytes();
+      } else {
+        size = 0;
+      }
+      this.callback = callback;
+    }
+
+    /**
+     * Marks a chunk as handled and invokes the callback and future that accompanied this chunk of data. Once a chunk is
+     * resolved, the data inside it is considered void.
+     * @param exception the reason for chunk handling failure.
+     */
+    private void resolveChunk(Exception exception) {
+      if (buf != null) {
+        future.done(exception == null ? size : 0, exception);
+        if (callback != null) {
+          callback.onCompletion(exception == null ? size : 0, exception);
+        }
+        buf = null;
+      }
+    }
+
+    private Callback<Long> makeCallbackForReadableStreamChannel(boolean primary, AtomicLong totalBytesRead) {
+      return (result, exception) -> {
+        lock.lock();
+        try {
+          if (exception == null) {
+            totalBytesRead.getAndAdd(result);
+          }
```

Comment:
only need to add once for primary.

Reply:
I was wondering if `totalBytesRead` for primary and secondary may not match in case we close secondary. Due to that, I was thinking to increment it for both. What do you say?